### DISCUSSION TITLE
Unknown arg when creating with many-to-many

### DISCUSSION THREAD

**Discussion by davatron5000** [prisma/prisma#4595] [Q&A] [answered]
👋    Howdy! I'm trying to create a many-to-many association when creating an item. I've made a reduced test case and things are working fine, but when I try to roll this into the main project, I get a weird error. 

The many-to-many data schema looks like this:

```prisma
datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URL")
}

generator client {
  provider = "prisma-client-js"
  previewFeatures = ["uncheckedScalarInputs"]
}

model Post {
  id         String      @id @default(uuid())
  title     String
  blocks    Block[]
}

model Block {
  id         String      @id @default(uuid())
  content String
  posts Post[]
}
```

I create two posts, then try to connect them to the UUID

```js
  await prisma.block.create({
    data: {
      content: "This is a block",
      posts: {
        connect: [{ id: '5d6442ee-9c5d-4044-97b8-6244a84d941f' },
                  { id: 'e7a54486-91eb-4123-9c6e-22bd175d3078' }]
      }
    }
  })

  const block = await prisma.block.findFirst({
    include: { posts: true }
  })
  console.log(block)
```

Like I said, this all works in the reduced test case, but when I roll it into the main project (that uses serverless functions on netlify inside a Nuxt app), I get the following error:

```
PrismaClientValidationError: 
Invalid `prisma.block.create()` invocation:

{
  data: {
    posts: {
    ~~~~~
      connect: [
        {
          id: 'cc96eaac-53b5-48ef-a73d-41b3347e37da'
        }
      ]
    },
    title: 'Testing',
  }
}

Unknown arg `posts` in data.posts for type BlockUncheckedCreateInput. Did you mean `id`? Available args:
type BlockUncheckedCreateInput {
  id?: String
  createdAt?: DateTime
  updatedAt?: DateTime
  isFeatured?: Boolean
  isActive?: Boolean
  title: String
  ownerId?: String | Null
  typeId?: String | Null
}
``` 

It's like the `BlockUncheckedCreateInput` (scalar?) type doesn't know about `posts` even though the many-to-many is modeled in the schema. Any ideas on what could be causing this or how to get around it?

Any input would be great. Been stuck on this for over a week. Thanks.

**Comment by ryands17** [comment]
Hey @davatron5000 👋 
Is `prisma generate` being run there on Netlify? It seems that this is a generated types issue. Also another way to check would be keeping your `@prisma/client` and `@prisma/cli` in sync with the same version (possibly the latest _2.13.0_) and deploying it again.

Let me know if this works.

**Comment by davatron5000** [comment, marked as answer]
@ryands17 I do have `prisma generate` running on Netlify, but I should have clarified this problem is happening locally with `netlify-cli` and `netlify dev`. I looked in to see if they do any kind of caching and it doesn't seem like they do.

But it does sound like the client isn't being recompiled or something. Hm. 🤔

**Reply by ryands17** [reply]
@davatron5000 As you mentioned, it's working fine locally so I assume it has to be the case where the client isn't being generated there properly. We also have E2E tests for platforms including Netlify [here](https://github.com/prisma/e2e-tests/tree/dev/platforms), so it should work without any issues. If it doesn't, I would suggest opening an issue [here](https://github.com/prisma/prisma/issues/new?assignees=&labels=&template=bug_report.md&title=) with the reproduction steps :)

**Reply by davatron5000** [reply]
Ok. After many hours of bit-by-bit trying to replicate the error in a reduced test case, I think i have a failing version (eliminating Netlify functions entirely). It seems the client is rejecting UUIDs (strings) when using relational scalars. 

This is all based on the `prisma-starter`project:

```schema.prisma
// schema.prisma

datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URL")
}

generator client {
  provider        = "prisma-client-js"
  previewFeatures = ["uncheckedScalarInputs"]
}

model Post {
  id     String  @id @default(uuid())
  title  String
  blocks Block[]
}

model Block {
  id      String  @id @default(uuid())
  content String
  posts   Post[]
  tag     Tag?    @relation(fields: [tagId], references: [id])
  tagId   String?
}

model Tag {
  id    String  @id @default(uuid())
  name  String  @unique
  Block Block[]
}
```

And then the script looks like this:

```ts
// script.ts

import { PrismaClient } from '@prisma/client'

const prisma = new PrismaClient()

// A `main` function so that you can use async/await
async function main() {
  const post1 = await prisma.post.create({
    data: {
      title: 'Hello'

    }
  });
  console.log(post1)
  const post2 = await prisma.post.create({
    data: {
      title: 'Hello Again',
    }
  })
  console.log(post2);

  const tag = await prisma.tag.create({
    data: {
      name: 'Test Tag',
    }
  })
  console.log(tag);

  const block = await prisma.block.create({
    data: {
      content: "This is a block",
      tagId: tag.id,
      posts: {
        connect: [{ id: post1.id },
                  { id: post2.id }]
      }
    }
  })
  console.log(block)

  const block1 = await prisma.block.findFirst({
    include: { posts: true }
  })
  console.log(block1)
}

main()
  .catch(e => {
    throw e
  })
  .finally(async () => {
    await prisma.$disconnect()
  })
```

Typescript fails to compile due at the `prisma.block.create` function with the same `BlockUncheckedCreateInput` I was getting in the OP with the JavaScript function (I probably should have showed the tagging stuff)

```console
TSError: ⨯ Unable to compile TypeScript:
script.ts:31:7 - error TS2322: Type '{ content: string; tagId: string; }' is not assignable to type '(Without<BlockCreateInput, BlockUncheckedCreateInput> & BlockUncheckedCreateInput) | (Without<...> & BlockCreateInput)'.
  Object literal may only specify known properties, and 'tagId' does not exist in type '(Without<BlockCreateInput, BlockUncheckedCreateInput> & BlockUncheckedCreateInput) | (Without<...> & BlockCreateInput)'.

31       tagId: tag.id,
```

If you have a quick answer for this, that would be rad. But if you'd like me to file a more formal issue, I totally can.

**Reply by ryands17** [reply]
Hey @davatron5000 👋 
This error is expected as you cannot use a combination of `uncheckedScalars` and `connect`. You would need to change your block create part to this instead :)

```ts
  const block = await prisma.block.create({
    data: {
      content: 'This is a block',
      tag: { connect: { id: tag.id } },
      posts: {
        connect: [{ id: post1.id }, { id: post2.id }],
      },
    },
})
```

Unchecked scalars can only be used individually and cannot be combined with a `connect`. Hope that makes sense :)

**Reply by Iamrahul4u** [reply]
you need to define unique constraints 

model Post {
  id         String      @id @default(uuid())
  title     String
  blocks    Block[]
}

model Block {
  id         String      @id @default(uuid())
  content String
  posts Post[]
  @@unique([content,id])
}

**Comment by CristhianMolinaWaldrop** [comment]
I have exactly the same error

**Comment by Qodestackr** [comment]
Anyone to provide a better way..?
Mine says:
```bash
Unknown arg `posts` in data.posts for type UserCreateInput. Did you mean `Post`? Available args:
type UserCreateInput {
  name?: String | Null
  email: String
  Post?: PostCreateNestedManyWithoutUserInput
  Profile?: ProfileCreateNestedOneWithoutUserInput
}
Unknown arg `profile` in data.profile for type UserCreateInput. Did you mean `Profile`? Available args:
type UserCreateInput {
  name?: String | Null
  email: String
  Post?: PostCreateNestedManyWithoutUserInput
  Profile?: ProfileCreateNestedOneWithoutUserInput
}
```

**Comment by joshribakoff-sm** [comment]
I got the same error, and it was due to passing Prisma invalid inputs:

```
    industry: {
    ~~~~~~~~
      connect: {
        id: NaN
      }
    },
```

**Comment by psygo** [comment]
On Next.js, I had to recompile things with `next dev` or `next build`.

**Comment by tuanpq1998** [comment]
In my case, I have changed `createMany` to `create` with `for...` loop to avoid this error, hope it helps someone

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
